Client/Frontend/Browser/BrowserViewController.swift

```diff
@@ -796,22 +798,45 @@ extension BrowserViewController: WKNavigationDelegate {
     }
 
     func webView(webView: WKWebView, didFinishNavigation navigation: WKNavigation!) {
+        let tab: Browser! = tabManager.getTab(webView)
+        if tab == nil {
+            return
+        }
+
+        tab.expireSnackbars()
+
         let notificationCenter = NSNotificationCenter.defaultCenter()
         var info = [NSObject: AnyObject]()
-        info["url"] = webView.URL
-        info["title"] = webView.title
+        info["url"] = tab.displayURL
+        info["title"] = tab.title
         notificationCenter.postNotificationName("LocationChange", object: self, userInfo: info)
 
-        if let tab = tabManager.getTab(webView) {
-            tab.expireSnackbars()
+        if let url = webView.URL?.absoluteString {
+            // The screenshot immediately after didFinishNavigation is actually a screenshot of the
+            // previous page, presumably due to some iOS bug. Adding a small delay seems to fix this,
+            // and the current page gets captured as expected.
+            let time = dispatch_time(DISPATCH_TIME_NOW, Int64(100 * NSEC_PER_MSEC))
+            dispatch_after(time, dispatch_get_main_queue()) {
+                if webView.URL?.absoluteString != url {
+                    // The page changed during the delay, so we missed our chance to get a thumbnail.
+                    return
+                }
+
+                if let screenshot = tab.screenshot(aspectRatio: ThumbnailCellUX.ImageAspectRatio, quality: 0.5) {
+                    let thumbnail = Thumbnail(image: screenshot)
+                    self.profile.thumbnails.set(webView.URL!, thumbnail: thumbnail, complete: nil)
```

Comment:
Yeah, but `url` is a string, and `thumbnails.set` needs an `NSURL`. The reason `url` is a string and not an `NSURL` is because I didn't want to hold a reference to the URL object (or else the "URL changed" check would never fail since the URL would always be equal to itself).